s2/gen.go

```diff
@@ -0,0 +1,1590 @@
+		// for best chance of table alignment.
+		extraStack = 32
+		allocStack = baseStack + extraStack + tableSize
+	)
+
+	// Memzero needs at least 128 bytes.
+	if tableSize < 128 {
+		panic("tableSize must be at least 128 bytes")
+	}
+
+	lenSrcBasic, err := Param("src").Len().Resolve()
+	if err != nil {
+		panic(err)
+	}
+	lenSrcQ := lenSrcBasic.Addr
+
+	stack := AllocLocal(allocStack)
+	table := stack.Offset(allocStack - tableSize)
+
+	tmpStack := baseStack
+	// Bail if we can't compress to at least this.
+	dstLimitPtrQ := stack.Offset(tmpStack)
+	tmpStack += 8
+	// dstStartPtrQ contains the original dst pointer for returning the length
+	dstStartPtrQ := stack.Offset(tmpStack)
+	tmpStack += 8
+	// sLimitL is when to stop looking for offset/length copies.
+	sLimitL := stack.Offset(tmpStack)
+	tmpStack += 4
+	// nextEmitL keeps track of the point we have emitted to.
+	nextEmitL := stack.Offset(tmpStack)
+	tmpStack += 4
+	// Repeat stores the last match offset.
+	repeatL := stack.Offset(tmpStack)
+	tmpStack += 4
+	// nextSTempL keeps nextS while other functions are being called.
+	nextSTempL := stack.Offset(tmpStack)
+	tmpStack += 4
+	// Ensure we have the correct extra stack.
+	// Could be automatic, but whatever.
+	if tmpStack-baseStack != extraStack {
+		log.Fatal("adjust extraStack to ", tmpStack-baseStack)
+	}
+
+	dstBaseBasic, err := Param("dst").Base().Resolve()
+	if err != nil {
+		panic(err)
+	}
+	dstBase := dstBaseBasic.Addr
+
+	if tmpStack > extraStack+baseStack {
+		panic(fmt.Sprintf("tmp stack exceeded", tmpStack))
+	}
+
+	// Zero table
+	{
+		iReg := GP64()
+		MOVQ(U32(tableSize/8/16), iReg)
+		tablePtr := GP64()
+		LEAQ(table, tablePtr)
+		zeroXmm := XMM()
+		PXOR(zeroXmm, zeroXmm)
+
+		Label("zero_loop_" + name)
+		for i := 0; i < 8; i++ {
+			MOVOU(zeroXmm, Mem{Base: tablePtr, Disp: i * 16})
+		}
+		ADDQ(U8(16*8), tablePtr)
+		DECQ(iReg)
+		JNZ(LabelRef("zero_loop_" + name))
+
+		// nextEmit is offset n src where the next emitLiteral should start from.
+		MOVL(iReg.As32(), nextEmitL)
+	}
+
+	{
+		const inputMargin = 8
+		tmp, tmp2, tmp3 := GP64(), GP64(), GP64()
+		MOVQ(lenSrcQ, tmp)
+		LEAQ(Mem{Base: tmp, Disp: -5}, tmp2)
+		// sLimitL := len(src) - inputMargin
+		LEAQ(Mem{Base: tmp, Disp: -inputMargin}, tmp3)
+		// dstLimit := len(src) - len(src)>>5 - 5
+		SHRQ(U8(5), tmp)
+		SUBL(tmp.As32(), tmp2.As32()) // tmp2 = tmp2 - tmp
+		MOVL(tmp3.As32(), sLimitL)
+		dstAddr := GP64()
+		MOVQ(dstBase, dstAddr)
+		// Store dst start address
+		MOVQ(dstAddr, dstStartPtrQ)
+		LEAQ(Mem{Base: dstAddr, Index: tmp2, Scale: 1}, tmp2)
+		MOVQ(tmp2, dstLimitPtrQ)
+	}
+
+	// s = 1
+	s := GP32()
+	MOVL(U32(1), s)
+	// repeatL = 1
+	MOVL(s, repeatL)
+
+	src := GP64()
+	Load(Param("src").Base(), src)
+
+	// Load cv
+	Label("search_loop_" + name)
+	candidate := GP32()
+	{
+		cv := GP64()
+		MOVQ(Mem{Base: src, Index: s, Scale: 1}, cv)
+		nextS := GP32()
+		// nextS := s + (s-nextEmit)>>6 + 4
+		{
+			tmp := GP64()
```

Comment:
I think this `tmp` is only used with `As32()`? If so you could use `GP32()` instead?